[ONNX] Add Support on 0d tensor Broadcast

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87212
* __->__ #87211

I am not sure if this will break things ...

Although 0d tensor is an undefined behavior in ONNX spec, I did some experiments and found that ONNX shape inference actually provides 0d as inference from 0d and 1d Op calculations, and the bug happened in Broadcast function. But still, if this breaks things really bad, I think we can put 0d tensor handling on hold, as it's not very common usage on models?

titaiwangms committed:
[ONNX] Add Support on 0d tensor Broadcast

[ghstack-poisoned]

AllenTiTaiWang commented:
cc @jcwchen

jcwchen commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -703,15 +703,21 @@ std::vector<::c10::ShapeSymbol> Broadcast(
+      // condition for corner case of 0d tensor
+      // 0d tensor with 1d tensor would give us 0d tensor
+      if (std::min(static_0_sz, static_1_sz) == 0) {
+        final_shape[rank_max - 1 - idx] = ::c10::ShapeSymbol::fromStaticSize(
```
nit: could we use shape_idx globally to represent `rank_max - 1 - idx` for simplicity?
```
size_t shape_idx = rank_max - 1 - idx;
```

AllenTiTaiWang replied:
Done

jcwchen reviewed (approved):
Thanks for fixing this. With this PR, PyTorch, Exporter, ONNX shape inference should be all align with the case (0d, 1d) -> 0d.

titaiwangms committed:
Update on "[ONNX] Add Support on 0d tensor Broadcast"


I am not sure if this will break things ...

Although 0d tensor is an undefined behavior in ONNX spec, I did some experiments and found that ONNX shape inference actually provides 0d as inference from 0d and 1d Op calculations, and the bug happened in Broadcast function. But still, if this breaks things really bad, I think we can put 0d tensor handling on hold, as it's not very common usage on models?

[ghstack-poisoned]

AllenTiTaiWang commented:
@BowenBao PTAL

BowenBao reviewed (approved):
LGTM. If you want, you can add a unit test in `test_pytorch_onnx_shape_inference.py`

AllenTiTaiWang commented:
@pytorchbot merge -g
